File: src/main/java/jm/app/service/OutlierDetector.java
package jm.app.service;

import jm.app.algebra.AlgebraUtil;
import jm.app.algebra.Matrix;
import jm.app.algorithm.GmmModel;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class OutlierDetector {

    private GmmModel gmmModel = new GmmModel();

    // the threshold of outlier
    private final double EPSILON = 0.001;

    public Map<String, Double[]> detectOutlier(Map<String, Double[]> records) {
        Map<String, Double[]> outliers = new HashMap<>();
        Map<Integer, String> indexToKey = new HashMap<>();
        // Step 1. Convert Map to Matrix
        Matrix x = null;
        Set<String> keys = records.keySet();
        if (null != keys && !keys.isEmpty()) {
            // get the number of features
            String firstKey = keys.iterator().next();
            Double[] firstFeatures = records.get(firstKey);
            int featureNum = firstFeatures.length;
            int recordNum = keys.size();
            x = new Matrix(recordNum, featureNum);
            // copy the data and record the index of each row
            int i = 0;
            for (String key : keys) {
                Double[] features = records.get(key);
                Matrix xi = new Matrix(1, featureNum);
                for (int c = 0; c < featureNum; c++) {
                    xi.setValue(0, c, features[c]);
                }
                AlgebraUtil.setRowVector(x, i, xi);
                indexToKey.put(i, key);
                i++;
            }
        } else {
            return null;
        }
        // Step 2. Calculate the mean vector u and the covariance matrix sigma
        x = AlgebraUtil.normalize(x, 0);
        Matrix u = AlgebraUtil.mean(x, 0);
        Matrix sigma = AlgebraUtil.covariance(x);

        // Step 3. Calculate the result of Gaussian Function and detect the outliers
        for (int i = 0; i < x.getRowNum(); i++) {
            Matrix xi = AlgebraUtil.getRowVector(x, i);
            BigDecimal p = gmmModel.gaussianFunction(xi, u, sigma);
            if (p.doubleValue() < EPSILON) {
                String key = indexToKey.get(i);
                outliers.put(key, records.get(key));
            }
        }
        return outliers;
    }

}

File: py/service/outlier_detector.py
import numpy as np
import py.algorithm.gmm_model as gmm_model
import py.algorithm.pca_model as pca
import py.algebra.albebra_util as au

class OutlierDetector():

    def __init__(self):
        self.max_feature_num = 2
        self.pca_model = pca.PcaModel()
        self.epsilon = 0.01

    def detect_outlier(self, records):
        key_set = list(records.keys())
        feature_num = len(records[key_set[0]])
        record_num = len(key_set)
        # Step 1. Convert Map to Matrix
        x = np.zeros((record_num, feature_num))
        for i in range(record_num):
            key = key_set[i]
            x[i, :] = records[key]

        # reduce dimension with PCA model if there are too many features
        if feature_num > self.max_feature_num:
            normalized_x = au.normalize(x, 0)
            x = np.zeros((record_num, self.max_feature_num))
            inputs = []
            for i in range(record_num):
                input = np.zeros((feature_num, 1))
                input[:, 0] = np.transpose(normalized_x[i, :])
                inputs.append(input)
            self.pca_model.train(inputs, self.max_feature_num)
            for i in range(record_num):
                xi = inputs[i]
                x[i, :] = np.transpose(self.pca_model.encode(xi))

        # Step 2. Calculate the mean vector u and the covariance matrix sigma
        x = au.normalize(x, 0)
        u = au.mean(x, 0)
        sigma = np.cov(np.transpose(x))

        # Step 3. Calculate the result of Gaussian Function and detect the outliers
        outputs = {}
        for i in range(record_num):
            xi = np.transpose(x[i, :])
            p = gmm_model.gaussian_fun(xi, u, sigma)
            if (p < self.epsilon):
                outputs[i] = xi
        return outputs


